src/dataflow/src/render/upsert.rs

```diff
@@ -398,18 +401,67 @@ where
                                     // We store errors as well as non-None values, so that they can be
                                     // retracted if new rows show up for the same key.
                                     let new_value = decoded_value.transpose();
+
+                                    use dataflow_types::plan::Permutation;
+                                    let new_value = new_value.map(|actual_value| {
+                                        actual_value.map(|non_err_val| {
+                                            let total_len = non_err_val.iter().count();
+                                            let (permutation, thinning) =
+                                                Permutation::construct_from_columns(
+                                                    &(0..decoded_key.iter().count())
+                                                        .into_iter()
+                                                        .collect::<Vec<usize>>(),
+                                                    total_len,
+                                                );
+                                            row_packer.clear();
+
+                                            row_packer.extend(
+                                                non_err_val
+                                                    .iter()
+                                                    .enumerate()
+                                                    .filter(|(i, _)| thinning.contains(i))
+                                                    .map(|(_, d)| d),
+                                            );
+                                            let r = row_packer.finish_and_reuse();
+                                            (r, permutation)
+                                        })
+                                    });
+
                                     let old_value = if let Some(new_value) = &new_value {
-                                        current_values.insert(decoded_key, new_value.clone())
+                                        current_values
+                                            .insert(decoded_key.clone(), new_value.clone())
                                     } else {
                                         current_values.remove(&decoded_key)
                                     };
+
+                                    // This closure re-uses `row_packer` to re-assemble rows with
+                                    // the keys that were replaced with `Datum::Dummy`
```

Comment:
This references `Datum::Dummy` but I'm not seeing it elsewhere in the code. Is it perhaps outdated?